---
title: "Basics Single Cell RNA-Seq Part 1: Create Seurat object"
author: "UCSF CAT Bioinformatics"
date: "`r Sys.Date()`"
output:
    html_document:
      keep_md: TRUE
      toc: TRUE
---

# Introduction to Single Cell RNA-Seq Part 1: Create Seurat object
```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, error = FALSE)
```

Our first Markdown analysis document concentrates on getting data into R and setting up our initial object. We will also replicate some of the tables and figures found in the Cellranger web summary.

### Load packages
We will start each section by loading the libraries necessary for that portion of the analysis.
```{r libraries}
library(Seurat)     # single cell RNA-Seq analysis
library(kableExtra) # format tables
library(ggplot2)   # create graphics
library(viridis)   # accessible color palettes
```

## Experiment metadata
The 10X data and metadata we have available for this subset of a recent experiment conducted at UCSF of LRTI (Lower Respiratory Tack Infection) sample during this workshop is very basic; we don't have any batches or conditions. I have also modified the read counts a little bit, so its essentially fake data.

### Setup the experiment folder and data info
```{r exp_setup, warning=FALSE,error=FALSE, message=FALSE, echo=FALSE}
experiment_name = "Workshop scRNA Seq UCSF"
dataset_loc <- "./01-CellRanger2"
experiment.metadata <- data.frame(ids = c("LRTI_WRK1", "LRTI_WRK2", "LRTI_WRK3", "LRTI_WRK4"),
                                  group = c("LRTI_Patient", "LRTI_Patient", c("LRTI_Patient", "LRTI_Patient")),
                                  run = c("Batch1", "Batch1","Batch2", "Batch2"))
experiment.metadata %>%
  kable() %>%
  kable_styling("striped")
```

## Cell Ranger sample metrics
```{r read_metrics, warning=FALSE,error=FALSE, message=FALSE, echo=FALSE}
g10x.metrics <- lapply(experiment.metadata$ids, function(i){
  metrics <- read.csv(file.path(dataset_loc,paste0(i,"_GXP"),"outs", "metrics_summary.csv"), colClasses = "character")
})
experiment.metrics <- do.call("rbind", g10x.metrics)
rownames(experiment.metrics) <- gsub("-","*",experiment.metadata$ids)

gxp_sequencing_metrics <- data.frame(t(experiment.metrics))
row.names(gxp_sequencing_metrics) <- gsub("\\."," ", rownames(gxp_sequencing_metrics))
```

This roughly replicates the table that appears in the Cell Ranger web summary file.
```{r gxp_table_metrics, warning=FALSE,error=FALSE, message=FALSE, echo=FALSE}
gxp_sequencing_metrics %>%
  kable(caption = 'GXP Cell Ranger Results') %>%
  pack_rows("Overview", 1, 3, label_row_css = "background-color: #666; color: #fff;") %>%
  pack_rows("Sequencing Characteristics", 4, 9, label_row_css = "background-color: #666; color: #fff;") %>%
  pack_rows("Mapping Characteristics", 10, 19, label_row_css = "background-color: #666; color: #fff;") %>%
  kable_styling("striped")
```

## Create Seurat object
We will be using [Seurat](http://satijalab.org/seurat/) as the basis of our single cell (or nucleus) RNA-Seq analysis. Seurat is a popular R package that is designed for QC, analysis, and exploration of single cell data, which aims to enable users to identify and interpret sources of heterogeneity from single cell transcriptomic measurements, and to integrate diverse types of single cell data.
In addition to the standard Seurat workflow, this documentation makes use of some custom code, and brings in functions from other packages. For additional information on Seurat standard workflows, see the authors' [tutorials](https://satijalab.org/seurat/vignettes.html).

### Read in expression matrix
First, we read in data from each individual sample folder.
```{r load_data_hdf5}
expression.data <- lapply(experiment.metadata$id, function(i){
  sample.matrix = Read10X_h5(file.path(dataset_loc, paste0(i,"_GXP"), "outs","raw_feature_bc_matrix.h5"))
  colnames(sample.matrix) = paste(sapply(strsplit(colnames(sample.matrix),split="-"), '[[', 1L), i, sep="+")
  sample.matrix
})
names(expression.data) <- experiment.metadata$id
```

```{r view_1, eval=FALSE}
View(expression.data)
```

### Merge matrices
```{r aggregate}
aggregate.data <- do.call("cbind", expression.data)
```

### Create Seurat Object

The `CreateSeuratObject` function allows feature (gene) and cell filtering by minimum cell and feature counts. We will set these to low numbers for now in order to explore manual filtering more fully in part 2.
```{r create_seurat_object}
gxp_experiment.aggregate <- CreateSeuratObject(
  aggregate.data,
  project = experiment_name,
  min.cells = 0,
  min.features = 300,
  names.field = 2, # tells Seurat which part of the cell identifier contains the sample name
  names.delim = "\\+")
```

## Add metadata
We can now attach the metadata in our table to the Seurat object.

### Match metadata to expression matrix
The columns of the expression matrix correspond to the cells in the experiment. When we created the Seurat object, the "names.field" and "names.delim" arguments allowed Seurat to infer sample identity from the cell names. This information is stored in a variable called "orig.ident."
```{r orig.ident}
levels(gxp_experiment.aggregate$orig.ident)
```

These sample identifiers are stored in the experiment.metadata object as well, which allows us to match the other metadata contained within that table to the correct cells within the Seurat object.
```{r sample.index}
sample.index <- match(gxp_experiment.aggregate$orig.ident, experiment.metadata$id)
```

### Attach metadata
The AddMetaData function returns a new Seurat object with an additional column in the metadata table containing the new information.
```{r AddMetaData}
gxp_experiment.aggregate[["group"]] <- factor(experiment.metadata$group[sample.index])

gxp_experiment.aggregate[["run"]] <- factor(experiment.metadata$run[sample.index])
```


### Add in percentage of reads that map to the mitochondrial genome

Filtering on the expression of genes from the mitochondrial genome is not appropriate in all cell types. However, in many tissues, low-quality / dying cells may exhibit extensive mitochondrial contamination. Even when not filtering on mitochondrial expression, the data can be interesting or informative.

The `PercentageFeatureSet` function calculates the proportion of counts originating from a set of features. Genes in the human mitochondrial genome begin with 'MT', while those in the mouse mitochondrial genome begin with 'mt'. These naming conventions make calculating percent mitochondrial very straightforward.

* Low-quality / dying cells often exhibit extensive mitochondrial contamination.
* We use the set of all genes, in human these genes can be identified as those that begin with 'MT'.

```{r add_mito, warning=FALSE,error=FALSE, message=FALSE, echo=FALSE}
gxp_experiment.aggregate[["percent.mito"]] <- PercentageFeatureSet(gxp_experiment.aggregate, pattern = "^MT-")
```

#### Add in percentage of reads that map to the ribosomal genes

Ribosomal genes also can also be highly represented, and can vary between cell types.
So its of interest to see how prevalent they are in the data.
We are selecting ribosomal protein genes rather than the actual rRNA, so theyâ€™re more a measure of the translational activity of the cell rather than the cleanliness of the polyA selection.

```{r}
gxp_experiment.aggregate[["percent.ribosomal"]] <- PercentageFeatureSet(gxp_experiment.aggregate, pattern = "^RP[LS]")
```


## Explore the Seurat object
A Seurat object is a complex data structure containing the data from a single cell or single nucleus assay and **all** of the information associated with the experiment, including annotations, analysis, and more. This data structure was developed by the authors of the Seurat analysis package, for use with their pipeline.
```{r view_2, eval=FALSE}
View(gxp_experiment.aggregate)
```

Most Seurat functions take the object as an argument, and return either a new Seurat object or a ggplot object (a visualization). As the analysis continues, more and more data will be added to the object.

```{r explore_object_1}
slotNames(gxp_experiment.aggregate)
gxp_experiment.aggregate@assays # a slot is accessed with the @ symbol
```
- Which slots are empty, and which contain data?
- What type of object is the content of the meta.data slot?
- What metadata is available?

There is often more than one way to interact with the information stored in each of a Seurat objects many slots. The default behaviors of different access functions are described in the help documentation.

```{r explore_object_2}
# which slot is being accessed here? find another way to produce the result
head(gxp_experiment.aggregate[[]]) %>%
  kable() %>%
  kable_styling("striped")
```

The use of syntax is often a matter of personal preference. In the interest of clarity, this documentation will generally use the more explicit syntax, with a few exceptions.

## Barcode inflection plots

Imagine the barcode rank plot from the Cell Ranger web summary. That graphic plots the number of UMIs against the barcode rank, and typically has a sharp inflection point where the number of UMIs drops dramatically. These points can represent a transition between cell types from a higher RNA content population to a lower RNA content population, or from cell-associated barcodes to background.

The Seurat `BarcodeInflectionsPlot` provides a similar graphic.

```{r barcode_inflection_plot}
gxp_experiment.aggregate <- CalculateBarcodeInflections(gxp_experiment.aggregate)
BarcodeInflectionsPlot(gxp_experiment.aggregate) +
  scale_color_viridis_d()
```

Adding a log-scale transformation to the x-axis increases the resemblance to the Cell Ranger plot. Values on the y-axis are already log-transformed.

```{r barcode_inflection_plot_log}
BarcodeInflectionsPlot(gxp_experiment.aggregate) +
  scale_x_continuous(trans = "log10") +
  scale_color_viridis_d()
```
```{r barcode_inflection_plot_values}
gxp_experiment.aggregate@tools$CalculateBarcodeInflections$inflection_points %>%
  kable() %>%
  kable_styling("striped")

SubsetByBarcodeInflections(object = gxp_experiment.aggregate)
```

Lets recreate the pretty Cell Ranger html plot


```{r fig_barcode_umi, warning=FALSE, error=FALSE, message=FALSE, echo=FALSE, fig.show="hold", out.width="50%"}
cr_filtered_cells <- as.numeric(gsub(",","",as.character(unlist(gxp_sequencing_metrics["Estimated Number of Cells",]))))

plot_cellranger_cells <- function(ind){
  xbreaks = c(1,1e1,1e2,1e3,1e4,1e5,1e6)
  xlabels = c("1","10","100","1000","10k","100K","1M")
  ybreaks = c(1,2,5,10,20,50,100,200,500,1000,2000,5000,10000,20000,50000,100000,200000,500000,1000000)
  ylabels = c("1","2","5","10","2","5","100","2","5","1000","2","5","10k","2","5","100K","2","5","1M")

  pl1 <- data.frame(
    index=seq.int(1,ncol(expression.data[[ind]])), nCount_RNA = sort(Matrix:::colSums(expression.data[[ind]])+1,decreasing=T),
    nFeature_RNA = sort(Matrix:::colSums(expression.data[[ind]]>0)+1,decreasing=T)
    ) %>% ggplot() + 
    scale_color_manual(values=c("grey50","red2","blue4"), labels=c("UMI_Background", "Features", "UMI_Cells"), name=NULL) +
    ggtitle(paste("CellRanger filltered cells:",experiment.metadata$ids[ind],sep=" ")) + xlab("Barcodes") + ylab("counts (UMI or Features") + 
    scale_x_continuous(trans = 'log2', breaks=xbreaks, labels = xlabels) + 
    scale_y_continuous(trans = 'log2', breaks=ybreaks, labels = ylabels) +
    geom_line(aes(x=index, y=nCount_RNA, color=index<=cr_filtered_cells[ind] , group=1), size=1.75) +
    geom_line(aes(x=index, y=nFeature_RNA, color="Features", group=1), size=1.25)

  return(pl1)
}

plot_cellranger_cells(1)
plot_cellranger_cells(2)
```

## Prepare for the next section

#### Save object
```{r saveRDS}
saveRDS(gxp_experiment.aggregate, file="scRNA_workshop-01.rds")
```

#### Download Rmd
```{r download_Rmd, eval=FALSE}
download.file("https://raw.githubusercontent.com/ucsf-cat-bioinformatics/2024-08-SCRNA-Seq-Analysis/main/data_analysis/02-filtering.Rmd", "02-filtering.Rmd")
```

#### Session information
```{r sessioinfo}
sessionInfo()
```
